content/post/reliable-change-detection.md

```diff
@@ -0,0 +1,254 @@
+
+By adding `Changed<C>` to our query filters, our query will only contain entities whose component of the type `C` has recently changed.
+This can be incredibly useful, allowing us to use **reactive** patterns in our code: while our systems may always run, they have minimal overhead and won't be doing wasted work.
+`Added<C>` query filters are also handy: they filter for components that were recently added instead, allowing you to automatically handle initialization in a later system.
+
+We can see how this plays out in practice by expanding our example above:
+
+```rust
+// TODO: more examples
+
+```
+
+## `DerefMut` magic
+
+Before we dive into the "reliable" part of reliable change detection, lets take a look at our basic building blocks.
+
+In Bevy, components (and resources) are marked as changed if a mutable reference to them has been unpacked, via a [custom implementation](https://github.com/bevyengine/bevy/blob/cf221f9659127427c99d621b76c8085c4860e2ef/crates/bevy_ecs/src/world/pointer.rs#L21) of the `DerefMut` trait on our `Mut` and `ResMut` smart pointers.
+
+This is fantastically clever: it allows us to quickly (and quite reliably) infer which pieces of data have been changed, without relying on caching the value and checking equality.
+While this can result in minor false positives (in case a component was mutably accessed but left unchanged), this can be easily avoided through simple and sane code quality.
+
+There are two quirks of this worth calling out:
+
+1. You can manually trigger change detection by manually mutably derefencing a piece of data, such as by calling `component.deref_mut()`.
+This is a very high-performance (if somewhat odd) way to flag components as requiring review, and even works on **marker components** (unit structs that store no data).
+2. Change detection fails to detect changes to data structures made via [interior mutability](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html).
+As above, this can either be a feature or a bug, depending on exactly what behavior you're trying to achieve.
+If you *really* wanted to, you could even combine these quirks, and call `deref_mut()` each time the data was changed as part of your setter methods to ensure change detection behaves as expected.
+Generally though, ECS prefers Plain-Old-Data in its components and resources, so this is unlikely to come up much in practice.
+
+## What does "reliable" mean?
+
+Those paying attention will notice a critical weasel word in my description above: "recently".
+This is at the crux of our issue: how long do we track these changes for.
+
+The obvious answer, and the one that Bevy used up to version 0.4 is that "changes persist until the end of the frame".
+This is simple, relatively easily implemented and high-performance.
+
+However, it also poses some serious usability problems:
+
+1. If the system that creates the change runs after the system that detects the change the change will be missed.
```

Comment:
Minor typo: you should probably add a comma in there between the two instances of "the change".

Reply:
```suggestion
1. If the system that creates the change runs after the system that detects the change, the change will be missed.
```